python/tank/platform/engine.py

```diff
@@ -1219,6 +1212,53 @@ def _do_invoke(self):
         # don't have ui so can't create an invoker!
         return None
 
+    def _generate_panel_id(self, app_instance_name, panel_name):
+        """
+        Compose a unique id for a panel. This is done based on the app instance
+        name plus the given panel name. By using the instance name rather than
+        the app name, we support the use case where more than one instance of an
+        app exists within a config.
+        :param app_instance_name: Instance name of the current application.
+        :param panel_name:        Name that uniquely identifies a panel.
+        :returns:                 Id to give to the panel.
+        """
+        panel_id = "%s_%s" % (app_instance_name, panel_name)
+        # to ensure the string is safe to use in most engines,
+        # sanitize to simple alpha-numeric form
+        panel_id = re.sub("\W", "_", panel_id)
+        panel_id = panel_id.lower()
+        return panel_id
+
+    def _setting_get_matching_panels(self, setting):
+        """
+        Reads a setting to find all the panels that match the given selectors.
+        Expects the following structure for the list in the setting:
+            {name: panel-name, app_instance: instance-name }
+        Use '' as a name to select all the panels of the instance-name.
+
+        :param setting: Name of the setting containing the panel selectors.
+        :returns:       A list of tuples for all panels that match the
+                        selectors. Each tuple is of the form:
+                          (instance-name, panel-id, callback)
+        """
+        return self.__setting_get_matching_items(setting, self.panels,
+                                                 self._generate_panel_id)
```

Comment:
I don't think we need this - let's chat about it.

Each panel app has got two separate steps that it will do (see the shogun panel as an example):

- register itself as a panel. This doesn't do anything or shows anything, it just tells the system about the fact that it is a panel.
- register a launch command. This is typically put on Shotgun menu that the user can click

So starting the panel (and every other UI) is always done via executing a *command*. Typically these commands are executed by a user clicking a menu, but we now do have a way to do it on startup too in flame.

What is `_setting_get_matching_panels` for? I think for the `tabs` setting in desktop, we can use `_setting_get_matching_commands` which will give you the command launch callback back and then execute that to actually show the panel. So I think the desktop code could be pretty similar to https://github.com/shotgunsoftware/tk-flame/blob/master/engine.py#L208 But maybe i am missing something ! :)